source/contribute.rst

```diff
@@ -41,128 +41,150 @@ contributing back to a Doctrine project.
    it will be `DBAL <https://github.com/doctrine/dbal>`_
--  Clone your fork locally
+-  Make the local repository aware of your fork.
 
 .. code-block:: console
 
-    $ git clone git@github.com:username/dbal.git
+    $ git clone --origin doctrine git@github.com:doctrine/dbal.git
 
--  Enter the dbal directory and add the **doctrine** remote
+At that point, you have a repository, and you should be able to start
+doing changes. Later on, if you end up with something you are satisfied
+with, you will need to be able to push your changes to your fork.
+
+Here is how you can make your local repository aware of your fork:
 
 .. code-block:: console
 
     $ cd dbal
-    $ git remote add doctrine git://github.com/doctrine/dbal.git
+    $ git remote add fork git@github.com:username/dbal.git
 
-Branching from the default branch
----------------------------------
+You should now have two remotes: one called ``doctrine``, another called
+``fork``.
 
-New pull requests are created with the repository's default branch as base branch.
-The default branch is the branch you see when you enter the repository page on GitHub.
+Choosing the right branch
+-------------------------
 
-.. image:: ../images/default-branch.png
-   :alt: The default branch
-   :style: margin-bottom: 20px
+Before you start working on a feature or a bug fix, you need to choose
+which branch you are going to contribute to. We try hard to follow
+semver, which means there are three choices.
 
-In this DBAL example, it's the branch with the name **2.11.x**. The branch name reflects the
-current lowest supported version of a repository.
+Let's say the latest version of the project is 2.3.1. The following
+branches should exist and are the possible choices:
 
-Newly introduced changes to 2.11.x will be up-merged at a later point in time
-to newer version branches (e.g. 2.12.x, 3.0.x). This way you don't have to
-re-introduce a new fix or feature of 2.11.x with another pull request to
-the other version branches.
+- ``2.3.x`` from which the next patch release will be tagged (2.3.2, 2.3.3, etc.);
+- ``2.4.x`` from which the next minor release will be tagged (2.4.0);
+- ``3.0.x`` from which the next major release will be tagged (3.0.0).
 
-Keeping the default branch up-to-date!
---------------------------------------
+Anything you contribute to a branch will be merged up to higher branches
+at some point. For example, if you contribute to ``2.3.x``, your changes
+will eventually land on ``2.4.x`` and ``3.0.x``.
 
-Once all this is done, you'll be able to keep your local branches up to
-date with the following command:
+The constraints we have are:
 
-.. code-block:: console
+- Every patch release should be more stable than the previous one.
+- An upgrade to a new minor version should not require any change on the
+  user side.
+- There should be a clear upgrade path from one major version to the
+  next.
+- We want to keep the branches as similar as possible to ease
+  maintenance.
```

Comment:
The general rules are as follows:
* Patch releases should be as stable as possible. They fix bugs, but in avoid making changes as far as possible.
* Improvements and additions are being made on the upcoming minor release branch. 
* An upgrade to a new minor version must not require any changes on the user's side.
* There must be a clear upgrade path from one major version to the next. That means necessary changes on the user's side will be advertised through deprecation notices, if anyway possible.
* Deprecations may be added for new minor versions only.
* Before a feature is removed (or an API changed) in a major version release, a corresponding deprecation notice must be included in a minor version release preceding it. 
* If a replacement feature and/or alternative API will be provided, it must be available in the minor release that adds the deprecation for the old feature.

And maybe one disputable extra item?

* When a minor release brings (major) changes in behavior, ideally those would be opt-in to avoid surprises during the upgrade process.

Reply:
> When a minor release brings (major) changes in behavior, ideally those would be opt-in to avoid surprises during the upgrade process.

That sounds like a breaking change, and wouldn't go in a minor release :thinking: 

I don't really understand what you mean I'm afraid.

Reply:
I notice that this is not editing the contributor's documentation, but instead making policies for the Doctrine project 🥶?

Reply:
That makes a lot more sense, yeah, let me add something about it.

Reply:
> I notice that this is not editing the contributor's documentation, but instead making policies for the Doctrine project cold_face?

Do you mean this feels out of place? Because https://www.doctrine-project.org/policies.html could be a good place for this.

Reply:
I mean, who am I to write or direct rules that mandate how the Doctrine shall organize/maintain its code 😐

Reply:
Dunno… a future maintainer hopefully :crossed_fingers: ?

Reply:
✅